pkg/apis/core/validation/validation.go

```diff
@@ -4551,12 +4604,52 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 		return allErrs
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		//TODO(vinaykul,InPlacePodVerticalScaling): Investigate if PodStatus.QOSClass in kubelet, elsewhere.
+		//    Once we can rely on Status persistence, drop this block. QOSClass does not change
+		//    once it is bootstrapped on podCreate. This needs to be addressed before beta.
+		//    Ref: https://github.com/kubernetes/kubernetes/pull/102884/#discussion_r663280487
+		// reject attempts to change pod qos
+		oldQoS := qos.GetPodQOS(oldPod)
+		newQoS := qos.GetPodQOS(newPod)
+		if newQoS != oldQoS {
+			allErrs = append(allErrs, field.Invalid(fldPath, newQoS, "Pod QoS is immutable"))
+		}
+	}
+
 	// handle updateable fields by munging those fields prior to deep equal comparison.
 	mungedPodSpec := *newPod.Spec.DeepCopy()
 	// munge spec.containers[*].image
 	var newContainers []core.Container
 	for ix, container := range mungedPodSpec.Containers {
 		container.Image = oldPod.Spec.Containers[ix].Image // +k8s:verify-mutation:reason=clone
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
```

Comment:
Can you comment here to the effect that this gate-check ensures that new in-place resize requests fail when then gate is off?

Reply:
Adding below comment:
```diff
diff --git a/pkg/apis/core/validation/validation.go b/pkg/apis/core/validation/validation.go
index 61d093ec9ff..39a965bd9d1 100644
--- a/pkg/apis/core/validation/validation.go
+++ b/pkg/apis/core/validation/validation.go
@@ -4623,6 +4623,8 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
        var newContainers []core.Container
        for ix, container := range mungedPodSpec.Containers {
                container.Image = oldPod.Spec.Containers[ix].Image // +k8s:verify-mutation:reason=clone
+               // When the feature-gate is turned off, any new requests attempting to update CPU or memory
+               // resource values will result in validation failure.
                if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
                        // Resources are mutable for CPU & memory only
                        //   - user can now modify Resources to express new desired Resources
```

Reply:
ACK